ktorm-support-postgresql/src/main/kotlin/org/ktorm/support/postgresql/BulkInsert.kt

```diff
@@ -223,3 +224,217 @@ public class BulkInsertOrUpdateOnConflictClauseBuilder : PostgreSqlAssignmentsBu
         )
     }
 }
+
+/**
+ * Bulk insert expression, represents a bulk insert statement in PostgreSQL.
+ *
+ * For example:
+ *
+ * ```sql
+ * insert into table (column1, column2) values (?, ?), (?, ?), (?, ?)...
+ * `on conflict (key) do update set (...) returning ...
+ * ```
+ *
+ * @property table the table to be inserted.
+ * @property assignments column assignments of the bulk insert statement.
+ * @property conflictColumns the index columns on which the conflict may happens.
+ * @property updateAssignments the updated column assignments while key conflict exists.
+ * @property returningColumns the columns to returning.
+ */
+public data class BulkInsertReturningExpression(
+    val table: TableExpression,
+    val assignments: List<List<ColumnAssignmentExpression<*>>>,
+    val conflictColumns: List<ColumnExpression<*>> = emptyList(),
+    val updateAssignments: List<ColumnAssignmentExpression<*>> = emptyList(),
+    val returningColumns: List<ColumnExpression<*>>,
+    override val isLeafNode: Boolean = false,
+    override val extraProperties: Map<String, Any> = emptyMap()
+) : SqlExpression()
+
+/**
+ * Construct a bulk insert expression in the given closure, then execute it and return the effected row count.
+ *
+ * The usage is almost the same as [batchInsert], but this function is implemented by generating a special SQL
+ * using PostgreSQL's bulk insert syntax, instead of based on JDBC batch operations. For this reason, its performance
+ * is much better than [batchInsert].
+ *
+ * The generated SQL is like: `insert into table (column1, column2) values (?, ?), (?, ?), (?, ?)...
+ * returning column1, column2, ...`.
+ *
+ * Usage:
+ *
+ * ```kotlin
+ * database.bulkInsertReturning(Employees) {
+ *     item {
+ *         set(it.name, "jerry")
+ *         set(it.job, "trainee")
+ *         set(it.managerId, 1)
+ *         set(it.hireDate, LocalDate.now())
+ *         set(it.salary, 50)
+ *         set(it.departmentId, 1)
+ *     }
+ *     item {
+ *         set(it.name, "linda")
+ *         set(it.job, "assistant")
+ *         set(it.managerId, 3)
+ *         set(it.hireDate, LocalDate.now())
+ *         set(it.salary, 100)
+ *         set(it.departmentId, 2)
+ *     }
+ *
+ *     returning (
+ *      it.id,
+ *      it.job
+ *     )
+ * }
+ * ```
+ *
+ * @since 3.4.0
+ * @param table the table to be inserted.
+ * @param block the DSL block, extension function of [BulkInsertStatementBuilder], used to construct the expression.
+ * @return the effected row count.
+ * @see batchInsert
+ */
+public fun <T : BaseTable<*>> Database.bulkInsertReturning(
+    table: T, block: BulkInsertReturningStatementBuilder<T>.(T) -> Unit
+): Pair<Int, CachedRowSet> {
```

Comment:
I don't think it is a good practice to return a `CachedRowSet`, as users have to obtain result values by column name, which is unsafe without strong type and static compiler checking. 

Is it possible to return a `List<TupleN<C1, C2>>`? It will be a great convenience as we can use the function like this: 

```kotlin
val results = database.bulkInsertReturning(Employees) {
    // ...
    returning(
        it.id,
        it.job
    )
}

for ((id, job) in results) {
    println(id)
    println(job)
}
```

BTW, if we just return one column `returning(it.id)`, the return type can be `List<C>`.

Reply:
Sorry @vincentlauvlwj I am confused on this one.

Do you have another function that implements this so I can see how you do it?

Reply:
I tried writing some pseudo code, the implementation might look like: 

```kotlin
fun <T : BaseTable<*>, R1 : Any, R2 : Any> Database.bulkInsertReturning(
    table: T, returning: Pair<Column<R1>, Column<R2>>, block: BulkInsertStatementBuilder<T>.(T) -> Unit
): List<Pair<R1?, R2?>> {
    val expression = ....
    val (_, rowSet) = executeUpdateAndRetrieveKeys(expression)
    val (r1, r2) = returning

    return rowSet.asIterable().map { row ->
        Pair(
            r1.sqlType.getResult(row, 1),
            r2.sqlType.getResult(row, 2)
        )
    }
}
```

Then the usage should be:

```kotlin
val listOfIdAndJob = database.bulkInsertReturning(Employees, returning = Pair(Employees.id, Employees.job)) {
    item {
        ....
    }
}

for ((id, job) in listOfIdAndJob) {
    println(id)
    println(job)
}
```

Reply:
But in this way, we need to add some overloaded version of `bulkInsertReturning` function. Maybe 3 is enough, they respectively return: 

- List<R?>
- List<Pair<R1?, R2?>>
- List<Triple<R1?, R2?, R3?>>

Reply:
Ok, I'll take a look into this, thank you for the examples.